sql/limit.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tamir Duberstein (tamird@gmail.com)
+
+package sql
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/cockroachdb/cockroach/sql/parser"
+)
+
+// limit constructs a limitNode based on the LIMIT and OFFSET clauses.
+func (*planner) limit(n *parser.Select, p planNode) (planNode, error) {
+	if n.Limit == nil {
+		return p, nil
+	}
+
+	var limit, offset int
+
+	data := []struct {
+		name       string
+		src        parser.Expr
+		dst        *int
+		defaultVal int
+	}{
+		{"LIMIT", n.Limit.Count, &limit, math.MaxInt64},
+		{"OFFSET", n.Limit.Offset, &offset, 0},
+	}
+
+	for _, datum := range data {
+		if datum.src == nil {
+			*datum.dst = datum.defaultVal
+		} else {
+			if parser.ContainsVars(datum.src) {
+				return nil, fmt.Errorf("argument of %s must not contain variables", datum.name)
+			}
+
+			normalized, err := parser.NormalizeExpr(datum.src)
+			if err != nil {
+				return nil, err
+			}
+			dstDatum, err := parser.EvalExpr(normalized)
+			if err != nil {
+				return nil, err
+			}
+
+			if dstDatum == parser.DNull {
+				*datum.dst = datum.defaultVal
+				continue
+			}
+
+			if dstDInt, ok := dstDatum.(parser.DInt); ok {
+				*datum.dst = int(dstDInt)
+				continue
+			}
+
+			return nil, fmt.Errorf("argument of %s must be type %s, not type %s", datum.name, parser.DummyInt.Type(), dstDatum.Type())
+		}
+	}
+
+	return &limitNode{planNode: p, limit: limit, offset: offset}, nil
+}
+
+type limitNode struct {
+	planNode
+	limit          int
+	offset         int
+	rowIndex       int
+	outputRowIndex int
+}
+
+func (n *limitNode) Next() bool {
+	if n.outputRowIndex >= n.limit {
+		return false
+	}
+
+	for n.rowIndex < n.offset && n.planNode.Next() {
+		n.rowIndex++
+	}
+
+	n.outputRowIndex++
+	return n.planNode.Next()
+}
+
+// What should this do?
+// func (n *limitNode) ExplainPlan() (name, description string, children []planNode) {
```

Comment:
I think it should return `"limit"` for the name and the count and offset for the description. Would be good to use `ALL` or some other indication if `count` is not set.